doc/nom_recipes.md

```diff
@@ -0,0 +1,235 @@
+# Nom Recipes
+
+These are short recipes for accomplishing common tasks with nom.
+
+* [Whitespace](#whitespace)
+  + [Wrapper combinators that eat whitespace before and after a parser](#wrapper-combinators-that-eat-whitespace-before-and-after-a-parser)
+* [Comments](#comments)
+  + [`// C++/EOL-style comments`](#----c---eol-style-comments-)
+  + [`/* C-style comments */`](#----c-style-comments----)
+* [Identifiers](#identifiers)
+  + [`Rust-Style Identifiers`](#-rust-style-identifiers-)
+* [Literal Values](#literal-values)
+  + [Escaped Strings](#escaped-strings)
+  + [Integers](#integers)
+    - [Hexadecimal](#hexadecimal)
+    - [Octal](#octal)
+    - [Binary](#binary)
+    - [Decimal](#decimal)
+  + [Floating Point Numbers](#floating-point-numbers)
+
+## Whitespace
+
+
+
+### Wrapper combinators that eat whitespace before and after a parser
+
+```rust
+/// A combinator that takes a parser `inner` and produces a parser that also consumes both leading and 
+/// trailing whitespace, returning the output of `inner`.
+fn ws<'a, F: 'a, O, E: ParseError<&'a str>>(inner: F) -> impl Fn(&'a str) -> IResult<&'a str, O, E>
+  where
+  F: Fn(&'a str) -> IResult<&'a str, O, E>,
```

Comment:
This can be `F: Parser<&'a str, O, E>` and the return type can be `impl Parser<&'a str, O, E>`

Reply:
The `Parser` trait is not publicly exported in v5.1.2.

Reply:
That's a good point; I didn't know that. You can remove the `'a` restriction on `F` if you don't pass `inner` by reference into `delimited()`. `F` can also be `FnMut(...) -> ...`.

Reply:
> You can remove the 'a restriction on F if you don't pass inner by reference into delimited(). 

There's no need to pass `inner` by reference, but if I elide the explicit lifetime `'a` I get a cryptic type error. [Here's a gist of my test dirty code](https://gist.github.com/rljacobson/840e99239eadc27b91d99feabf40a50a#file-nom-recipes-testing-rs-L25) if you want to take a look. (I wish the rust playground had the nom crate.)

> `F` can also be `FnMut(...) -> ...`.

Another version difference: `nom::sequence::delimited` requires an `Fn` in [v5.1.2](https://docs.rs/nom/5.1.2/nom/sequence/fn.delimited.html).

But since this is the second time it's come up, maybe the recipes *should* target v6.0.0. What do you think?

Reply:
Hmm, I just started using Nom, so I didn't know there were so many breaking changes. I don't think it matters so much which version the examples are for, but it would be good to point out any changes that need to be made to work with the other version.
As far as the lifetime is concerned, I don't see any type errors on your gist version. The lifetime is only required if you pass `inner` by reference, since `inner` will go out of scope at the end of the function (before any parsing is done).